usr/src/lib/libzfs/common/libzfs_crypto.c

```diff
@@ -0,0 +1,1293 @@
+ * (the "License").  You may not use this file except in compliance
+ * with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016, Datto, Inc. All rights reserved.
+ */
+
+#include <string.h>
+#include <strings.h>
+#include <ctype.h>
+#include <unistd.h>
+#include <sys/zfs_context.h>
+#include <sys/fs/zfs.h>
+#include <sys/dsl_crypt.h>
+#ifdef sun
+#include <kmfapi.h>
+#include <security/pkcs11.h>
+#include <cryptoutil.h>
+#else
+#include <sys/crypto/icp.h>
+#endif
+#include <libintl.h>
+#include <termios.h>
+#include <signal.h>
+#include <errno.h>
+#include <libzfs.h>
+#include "libzfs_impl.h"
+#include "zfeature_common.h"
+
+/*
+ * User keys are used to decrypt the master encyrption keys of a dataset. This
+ * indirection allows a user to change his / her access key without having to
+ * re-encrypt the entire dataset. User keys can be provided in one of several
+ * ways. Raw keys are simlply given to the kernel as is. Similarly, hex keys
+ * are converted to binary and passed into the kernel. Password based keys are
+ * a bit more complicated. Passwords alone do not provide suitable entropy for
+ * encryption and may be too short or too long to be used. In order to derive
+ * a more appropriate key we use a PBKDF2 function. This function is designed
+ * to take a (relatively) long time to calculate in order to discourage
+ * attackers from guessing from a list of common passwords. PBKDF2 requires
+ * 2 additional parameters. The first is the number of iterations to run, which
+ * will ultimately decide how long it takes to derive the resulting key from
+ * the password. The second parameter is a salt that is randomly generated for
+ * each datasset. The salt is used to "tweak" PBKDF2 such that a group of
+ * attackers cannot reasonably generate a table of commonly known passwords to
+ * their output keys and expect it work for all past and future PBKDF2 users.
+ * We store the salt as a hidden property of the dataset (although it is
+ * technically ok if the salt is known to the attacker).
+ */
+
+typedef enum key_format {
+	KEY_FORMAT_NONE = 0,
+	KEY_FORMAT_RAW,
+	KEY_FORMAT_HEX,
+	KEY_FORMAT_PASSPHRASE
+} key_format_t;
+
+typedef enum key_locator {
+	KEY_LOCATOR_NONE,
+	KEY_LOCATOR_PROMPT,
+	KEY_LOCATOR_URI
+} key_locator_t;
+
+#define	MIN_PASSPHRASE_LEN 8
+#define	MAX_PASSPHRASE_LEN 64
+#define	DEFAULT_PBKDF2_ITERATIONS 100000
```

Comment:
This has been upped to 350000 (minimum 100000) on ZoL. Those changes will be reflected here after some more review. On a modern i7 this takes about .4 s